src/Catalyst.Core.Modules.Keystore/KeyStoreService.cs

```diff
@@ -0,0 +1,830 @@
+            if (ekey == null)
+            {
+                throw new KeyNotFoundException($"The key '{keyName}' does not exist.");
+            }
+
+            AsymmetricCipherKeyPair kp = null;
+            UseEncryptedKey(ekey, key => { kp = GetKeyPairFromPrivateKey(key); });
+
+            // A signer for the key.
+            var ku = new KeyUsage(KeyUsage.DigitalSignature
+              | KeyUsage.DataEncipherment
+              | KeyUsage.KeyEncipherment);
+            ISignatureFactory signatureFactory = null;
+            switch (kp.Private)
+            {
+                case ECPrivateKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        X9ObjectIdentifiers.ECDsaWithSha256.ToString(),
+                        kp.Private);
+                    break;
+                case RsaPrivateCrtKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        PkcsObjectIdentifiers.Sha256WithRsaEncryption.ToString(),
+                        kp.Private);
+                    break;
+                case Ed25519PrivateKeyParameters _:
+                    signatureFactory = new Asn1SignatureFactory(
+                        EdECObjectIdentifiers.id_Ed25519.Id,
+                        kp.Private);
+                    ku = new KeyUsage(KeyUsage.DigitalSignature);
+                    break;
+            }
+
+            if (signatureFactory == null)
+            {
+                throw new NotSupportedException($"The key type {kp.Private.GetType().Name} is not supported.");
+            }
+
+            // Build the certificate.
+            var dn = new X509Name($"CN={ekey.Id}, OU=keystore, O=ipfs");
+            var ski = new SubjectKeyIdentifier(Base58.Decode(ekey.Id));
+
+            // Not a certificate authority.
+            // TODO: perhaps the "self" key is a CA and all other keys issued by it.
+            var bc = new BasicConstraints(false);
+
+            var certGenerator = new X509V3CertificateGenerator();
+            certGenerator.SetIssuerDN(dn);
+            certGenerator.SetSubjectDN(dn);
+            certGenerator.SetSerialNumber(BigInteger.ValueOf(1));
+            certGenerator.SetNotAfter(DateTime.UtcNow.AddYears(10));
+            certGenerator.SetNotBefore(DateTime.UtcNow);
+            certGenerator.SetPublicKey(kp.Public);
+            certGenerator.AddExtension(X509Extensions.SubjectKeyIdentifier, false, ski);
+            certGenerator.AddExtension(X509Extensions.BasicConstraints, true, bc);
+            certGenerator.AddExtension(X509Extensions.KeyUsage, false, ku);
+
+            return certGenerator.Generate(signatureFactory);
+        }
+
+        private sealed class PasswordFinder : IPasswordFinder, IDisposable
+        {
+            public char[] Password;
+
+            public void Dispose() { Password = null; }
+
+            public char[] GetPassword() { return Password; }
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused private method 'GetPassword'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)